File: day02/day02.py
from functools import reduce

data = open('input.txt', 'r').readlines()
#data = ["forward 5", "down 5", "forward 8", "up 3", "down 8", "forward 2"]

def apply_commands_part1(cmds):
    def apply_command(pos, cmd):
        [op, count] = cmd.split()
        if op == 'forward':
            return (pos[0]+int(count), pos[1])
        if op == 'down':
            return (pos[0], pos[1]+int(count))
        if op == 'up':
            return (pos[0], pos[1]-int(count))
        raise ValueError(f'invalid op {op}')
    return reduce(apply_command, cmds, (0, 0))

def apply_commands_part2(cmds):
    def apply_command(pos, cmd):
        [op, count] = cmd.split()
        if op == 'forward':
            return (pos[0]+int(count), pos[1]+pos[2]*int(count), pos[2])
        if op == 'down':
            return (pos[0], pos[1], pos[2]+int(count))
        if op == 'up':
            return (pos[0], pos[1], pos[2]-int(count))
        raise ValueError(f'invalid op {op}')
    return reduce(apply_command, cmds, (0, 0, 0))

pos = apply_commands_part1(data)
print(f'Final position, part 1: horizontal={pos[0]}, depth={pos[1]}, product={pos[0]*pos[1]}')
pos = apply_commands_part2(data)
print(f'Final position, part 2: horizontal={pos[0]}, depth={pos[1]}, product={pos[0]*pos[1]}')

File: day02/day02.rs
use std::error::Error;
use std::fmt::Display;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;

fn main() -> Result<(), Box<dyn Error>> {
    let f = File::open("input.txt")?;
    let mut reader = BufReader::new(f);
    let prog = parse(&mut reader)?;

    let p1 = run(&prog, apply1);
    println!("Final position, part 1: {}", p1);
    let p2 = run(&prog, apply2);
    println!("Final position: part 2: {}", p2);
    Ok(())
}

enum Op {
    Forward(i32),
    Up(i32),
    Down(i32),
}

impl Display for Op {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::result::Result<(), ::std::fmt::Error> {
        match *self {
            Op::Forward(n) => f.write_fmt(format_args!("Forward({})", n)),
            Op::Up(n) => f.write_fmt(format_args!("Up({})", n)),
            Op::Down(n) => f.write_fmt(format_args!("Down({})", n)),
        }
    }
}

fn parse<R: BufRead>(r: &mut R) -> Result<Vec<Op>, Box<dyn Error>> {
    let mut prog = Vec::new();
    for line in r.lines() {
        let l = line?;
        let i = l.find(' ').ok_or("no space in line")?;
        let count = l[i + 1..].parse::<i32>()?;
        let cmd = match &l[..i] {
            "forward" => Ok(Op::Forward(count)),
            "up" => Ok(Op::Up(count)),
            "down" => Ok(Op::Down(count)),
            _ => Err("invalid command"),
        }?;
        prog.push(cmd);
    }
    Ok(prog)
}

struct Pos {
    horizontal: i32,
    depth: i32,
    aim: i32,
}

impl Pos {
    fn new() -> Pos {
        Pos {
            horizontal: 0,
            depth: 0,
            aim: 0,
        }
    }
}

impl Display for Pos {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::result::Result<(), ::std::fmt::Error> {
        f.write_fmt(format_args!(
            "horizontal={}, depth={}, product={}",
            self.horizontal,
            self.depth,
            self.horizontal * self.depth
        ))
    }
}

fn run(prog: &Vec<Op>, apply: fn(Pos, &Op) -> Pos) -> Pos {
    prog.iter().fold(Pos::new(), apply)
}

fn apply1(p: Pos, cmd: &Op) -> Pos {
    match cmd {
        Op::Forward(n) => Pos {
            horizontal: p.horizontal + n,
            depth: p.depth,
            aim: p.aim,
        },
        Op::Down(n) => Pos {
            horizontal: p.horizontal,
            depth: p.depth + n,
            aim: p.aim,
        },
        Op::Up(n) => Pos {
            horizontal: p.horizontal,
            depth: p.depth - n,
            aim: p.aim,
        },
    }
}

fn apply2(p: Pos, cmd: &Op) -> Pos {
    match cmd {
        Op::Forward(n) => Pos {
            horizontal: p.horizontal + n,
            depth: p.depth + n * p.aim,
            aim: p.aim,
        },
        Op::Down(n) => Pos {
            horizontal: p.horizontal,
            depth: p.depth,
            aim: p.aim + n,
        },
        Op::Up(n) => Pos {
            horizontal: p.horizontal,
            depth: p.depth,
            aim: p.aim - n,
        },
    }
}
